### PR TITLE
Add `SharedAllocatorList`

### PR DESCRIPTION
**Pull request by 0xEAB**
Revamp of #6465; recreated by applying the changes to current ~master.

The unresolved reviews of the original PR haven’t been addressed yet.

### CODE REVIEW

File: std/experimental/allocator/building_blocks/allocator_list.d [hunk 1]
```diff
@@ -661,65 +666,271 @@ version (Posix) @system unittest
...
+    import std.typecons : Ternary;
+    import std.traits : hasMember;
+    import core.internal.spinlock : SpinLock;
+
+private:
+    // Forward all calls to 'impl' and protect them by the lock below
+    AllocatorList!(Factory, BookkeepingAllocator) impl;
+    SpinLock lock = SpinLock(SpinLock.Contention.brief);
+
+    // Remove this when 'https://github.com/dlang/druntime/pull/2156' gets merged
```
Comment (thewilsonator) [inline std/experimental/allocator/building_blocks/allocator_list.d:681]: does this have a corresponding dmd PR? is it still open?

Comment (0xEAB) [reply std/experimental/allocator/building_blocks/allocator_list.d:681]: I haven’t been able to find any successor to that druntime PR.

Comment (0xEAB) [reply std/experimental/allocator/building_blocks/allocator_list.d:681]: Should we rather rephrase that note or simply remove it?

Comment (0xEAB) [reply std/experimental/allocator/building_blocks/allocator_list.d:681]: Theoretically, there’s https://github.com/dlang/dmd/issues/17128.

Comment (0xEAB) [reply std/experimental/allocator/building_blocks/allocator_list.d]: So, I’ve replaced this comment in 45a094f57629f90e9f8bca82fc55b5ee4a6811aa.

File: std/experimental/allocator/building_blocks/allocator_list.d [hunk 2]
```diff
@@ -661,65 +666,271 @@ version (Posix) @system unittest
...
+    }
+
+    // Create an allocator based upon 4MB regions, fetched from the GC heap.
+    AllocatorList!((n) => BorrowedRegion!()(new ubyte[max(n, 1024 * 4096)])) a1;
+    SharedAllocatorList!((n) => SharedBorrowedRegion!()(new ubyte[max(n, 1024 * 4096)])) a2;
 
-    assertThrown!AssertError(a.alignedAllocate(size_t.max, 1024));
-    a.deallocateAll();
+    testAlloc(a1);
+    testAlloc(a2);
```
Comment (0xEAB) [inline std/experimental/allocator/building_blocks/allocator_list.d:933]: This call currently takes forever…

Comment (0xEAB) [reply std/experimental/allocator/building_blocks/allocator_list.d:938]: Bypassed.

**Review by thewilsonator — APPROVED** [review]
Not that you need do it here, but if this allocator is advertised as thread safe then the unit tests should not be system, we should `@trusted` the minimal interface and have the rest be safe.

**Comment by 0xEAB** [comment]
> we should @trusted the minimal interface and have the rest be safe.

I suppose this would need a purpose-built shim as [“allocators must be entirely `@system`”](https://github.com/dlang/phobos/pull/9078#issuecomment-2449067129).

**Comment by thewilsonator** [comment]
That only lists `free` as a function that must be `@system` because use after free, and double free.

**Pull request merged**